porch/repository/pkg/git/git.go

```diff
@@ -643,3 +630,106 @@ func (r *gitRepository) getRepo() (string, error) {
 
 	return origin.Config().URLs[0], nil
 }
+
+func (r *gitRepository) update(ctx context.Context) error {
+	auth, err := r.getAuthMethod(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Fetch
+	switch err := r.repo.Fetch(&gogit.FetchOptions{
+		RemoteName: originName,
+		Auth:       auth,
+		Tags:       gogit.AllTags,
+	}); err {
+	case nil: // OK
+	case gogit.NoErrAlreadyUpToDate:
+	case transport.ErrEmptyRemoteRepository:
+
+	default:
+		return fmt.Errorf("cannot fetch repository %s/%s: %w", r.namespace, r.name, err)
+	}
+
+	// Create tracking branches for remotes
+	refs, err := r.repo.References()
+	if err != nil {
+		return fmt.Errorf("cannot identify repository remote references in %s/%s: %w", r.namespace, r.name, err)
+	}
+
+	// Collect remote and local branches. Both maps are indexed by the local reference name.
+	// remote references (refs/remotes/origin/...) are transformed to `refs/heads/...` to have
+	// both maps use matching keys.
+	remoteBranches := map[string]*plumbing.Reference{}
+	localBranches := map[string]*plumbing.Reference{}
+	for {
+		ref, err := refs.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			klog.Warningf("Skipping reference during iteration on error: %v", err)
+			continue
+		}
+
+		name := ref.Name().String()
+		switch {
+		case strings.HasPrefix(name, refRemoteBranchPrefix):
+			branch := strings.TrimPrefix(name, refRemoteBranchPrefix)
+			remoteBranches[refHeadsPrefix+branch] = ref
+
+		case strings.HasPrefix(name, refHeadsPrefix):
+			localBranches[name] = ref
+		}
+	}
+
+	// TODO: Explore use of automatic branch tracking to do this.
+	// There may be risks involved such as automated rebase which may lead to incorrect results on the package.
+	// One possibility is to replay package history ... something to consider in the future.
+	for name, remoteRef := range remoteBranches {
+		localRef, ok := localBranches[name]
+
+		if !ok {
+			localRef = plumbing.NewHashReference(plumbing.ReferenceName(name), remoteRef.Hash())
+			// local branch doesn't exist. create it
+			if err := r.repo.Storer.SetReference(localRef); err != nil {
+				klog.Warningf("Skipping failed local ref %q: %v", name, err)
```

Comment:
Do we expect this?  Maybe just return an error and fail the whole update?

Reply:
sg.